src/doc/src/reference/semver.md

```diff
@@ -273,6 +275,58 @@ Mitigation strategies:
   a struct to prevent users from using struct literal syntax, and instead
   provide a constructor method and/or [Default] implementation.
 
+<a id="struct-change-repr-when-public"></a>
+### Major: change a from well-defined repr to another when all fields are public
+
+When a struct has a well-defined repr (`transparent`, `C`) and only public fields, this will break unsafe code relying on 
+that repr.
+
+Adding a well-defined repr to a `repr(Rust)` struct is *not* a breaking change.
```

Comment:
This *can* be a breaking change. It could change the struct layout in ways that violates assumptions. For instance, alignment, or total structure size.

Reply:
But you already can't assume anything about the alignment or size of a repr(rust) struct, so assuming that with a volatile (or anything else) is already illegal.

```rust
// this structure *might* have any alignment 8
struct Thing {
  x: i32
}
```

Reply:
"shouldn't" is not "can't". It is *possible* to assume the size or alignment of a struct, and it's *possible* for crates depending on yours to care about that, and therefore it's *possible* that for your crate it's a breaking change. It may or may not be one in practice.

Reply:
If you're relying on the layout of a repr(rust) type then a stable compiler upgrade is already allowed to break your code. Even a simple recompilation is allowed to break the code. The programmer has *at best* already accepted that they don't care about part of semver.

As a similar example, it's also "shouldn't" rather than "can't" for a lot of transmutation things that "usually don't actually break".
```rust
struct Foo{
  p: *const i32,
  l: usize
}
```
If some transmutes a `&[i32]` into `Foo`, should they be allowed to complain about a breaking change happening to them if it stops working how they expect? They *can* type it into a file and make the compiler compile it after all. And it *will* work on the current stable compiler. But if it doesn't work in the next stable, should they be allowed to complain about a semver break?

Reply:
While changing `#[repr(Rust)]` to `#[repr(C)]` might break code, that is still not a breaking (major) change in my opinion. As Lokathor pointed out, a stable ocmpiler upgrade could also break it. Relying on `#[repr(Rust)]` structs is not something that someone *should* do, and if you still do it, you have to live with the changes. Rust makes no guarantees about `#[repr(Rust)]`, so no assumptions should be made about it.

Reply:
It's like relying on `#[doc(hidden)]` items, or relying on a crate that has the version `0.0.1`. You can do it, but the library can also just break it in a patch.

Reply:
@joshtriplett, the SemVer guidelines here are guided by RFC 1105, and generally these guidelines do not consider *every* potentially breaking change as a major change. There are several that are allowed, and some that I categorized as "possibly-breaking" where it is a gray area. The introduction of this chapter tries to clarify how this guidance works, and it does try to highlight gray areas in various places.

The guidelines are based on a risk spectrum, and the goal here is to provide library authors with enough information so they can evaluate those risks, and figure out where changes for their project land on that spectrum (which may be different for different projects and different kinds of changes).

I think changing a `repr(Rust)` type is fine. Although users may rely on that, they shouldn't, and it is documented in several places that they shouldn't. 

Another example would be type inference failures, where this is a general risk of some changes that aren't necessarily considered breaking. 

I think it would be reasonable to leave some sidenotes in this section describing the potential risks of other `repr` changes that might potentially be breaking, but aren't usually considered major changes. These notes can try to make it clear that there is some risk here, and let the reader decide if those are reasonable.